agent/csi/volumes.go

```diff
@@ -0,0 +1,263 @@
+package csi
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+)
+
+// volumes is a map that keeps all the currently available volumes to the agent
+// mapped by volume ID.
+type volumes struct {
+	mu     sync.RWMutex // To sync between Add(), Remove() and Reset() for "m" map
+	lock   sync.RWMutex // To sync between Get(), addVolumes() and removeVolumes()
+	m      map[string]*api.VolumeAssignment
+	plugin *NodePlugin
+}
+
+const maxRetry float64 = 25
+
+const initialBackoff = 100 * time.Millisecond
+const factor float64 = 2
+
+// NewManager returns a place to store volumes.
+func NewManager() exec.VolumesManager {
+	return &volumes{
+		m:      make(map[string]*api.VolumeAssignment),
+		plugin: NewNodePlugin(),
+	}
+}
+
+// Get returns a volume published path for the provided volume ID.  If the volume doesn't exist, returns empty string.
+func (r *volumes) Get(volumeID string) (string, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return r.plugin.GetPublisedPath(volumeID)
+}
+
+// Add adds one or more volumes to the volume map.
+func (r *volumes) Add(volumes ...api.VolumeAssignment) {
+	r.mu.Lock()
+	var volumeObjects []*api.VolumeAssignment
+	defer r.mu.Unlock()
+	ctx := context.Background()
+	for _, volume := range volumes {
+		v := volume.Copy()
+		log.G(ctx).WithField("method", "(*volumes).Add").Debugf("Add Volume:%v", volume.VolumeID)
+
+		r.m[volume.VolumeID] = v
+		volumeObjects = append(volumeObjects, v)
+	}
+	go r.addVolumes(volumeObjects)
+}
+
+func (r *volumes) addVolumes(volumeObjects []*api.VolumeAssignment) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	ctx := context.Background()
+	var failedStage []*api.VolumeAssignment
+	var failedPublish []*api.VolumeAssignment
+	for _, v := range volumeObjects {
+		err := r.plugin.NodeStageVolume(ctx, v)
+		if err == nil {
+			log.G(ctx).WithField("method", "(*volumes).addVolumes").Debugf("Volume:%v staged successfully", v.VolumeID)
+			errPub := r.plugin.NodePublishVolume(ctx, v)
+			if errPub != nil {
+				failedPublish = append(failedPublish, v)
+			} else {
+				log.G(ctx).WithField("method", "(*volumes).addVolumes").Debugf("Volume:%v published successfully", v.VolumeID)
+			}
+		} else {
+			failedStage = append(failedStage, v)
+		}
+	}
+
+	if len(failedStage) > 0 {
+		r.retryAddVolumes(ctx, failedStage)
+	}
+
+	if len(failedPublish) > 0 {
+		r.retryPublishVolume(ctx, failedPublish)
+	}
+}
+
+func (r *volumes) retryAddVolumes(ctx context.Context, volumeObjects []*api.VolumeAssignment) {
```

Comment:
There is a lot of duplication between this and `addVolumes`. They should be condensed into a single method.